_posts/2020-09-21-centralised-schema-management-microservices.md

```diff
@@ -0,0 +1,209 @@
+
+CSM is a tool and architectural pattern to manage that problem. It forces all parties interested in a given kind of data to use a standard definition of it &mdash; what properties the data has, what's the set of allowed values for each property, which ones are mandatory, and so forth. This is normally not optimal from the viewpoint of any particular service, since depending on how they're built they usually have a richer set of tools available than the smallest common denominator an external schema represents. For example, a TypeScript-based service would rather use its own type system for defining objects, and a relational schema defined in SQL in the fourth normal form is a thing of beauty compared to any JSON schema document.
+
+For systems that have just a few kinds of data, or just a few different services that deal with it, the tradeoff is likely not worthwhile. But when you have a dozen services that all need to be able to process some or all of the set of 100 data types, implementing CSM is the only way to stave off the impending madness. Even if that results in all of your services having to submit to a jackbooted overlord who controls what they may or may not say in public. (Please do not try to extrapolate this into anything human-related.)
+
+# A portrait of a CSM in three acts
+
+## The MVP
+
+To be of any use, a CSM implementation, whether homebrew or off-the-shelf, needs to include at least the following:
+
+- __Some way to distribute schemas__ in a format or formats that are understood by all parts of the system.
+   - Your options format-wise include [JSON Schema](https://json-schema.org/), [Apache Avro](https://avro.apache.org/docs/current/spec.html), [Protobuf](https://developers.google.com/protocol-buffers/docs/proto) schemas, and the comedy options of XML Schema or RELAX-NG. (Not going to link to those two.)
+   - Ideally you define your schema just once, and distribute it in multiple formats.
+   - If you have, or suspect you'll have in the future any external APIs, then your selection narrows. There are not many formats that are widely accepted, and JSON Schema is probably the least problematic option.
+- For each of your services, a means of __consuming (using) the schema__. Having multiple formats available makes it easier to find a good consumer or validator.
+   - It's a nice bonus if your consumer can process the schema into something native for your under-the-hood implementation. This includes F# type providers, Clojure spec or Plumatic schema, or even just plain old Java code generators. (Let's not mention JAXB.)
```

Comment:
Personally I would say the most important thing is to be able to generate a client for an API. Otherwise the user will be forced to deal with the limited expressibility of the used _wire encoding_. For example, JSON doesn't have a representation for dates so they are wire encoded in either an integer or a string, and this is irrelevant and uninteresting to the actual users of the data. I think it's a bit wrong idea to "process into something native", since the important thing is to process away from the simplified wire encoding into something richer (native or not). What do you think?